services/observation_plan_queue/observation_plan_queue.py

```diff
@@ -32,16 +33,163 @@
+                "start_date": allocationsAndStartDate[0]["start_date"],
+            }
+        earliest = None
+        for allocationAndStartDate in allocationsAndStartDate:
+            # find the plan
+            if earliest is None:
+                earliest = allocationAndStartDate
+                continue
+            if (
+                telescopeAllocationLookup[allocationAndStartDate["allocation_id"]][
+                    "morning"
+                ]
+                is False
+            ):
+                continue
+            start_date = allocationAndStartDate["start_date"]
+            if start_date is None:
+                continue
+            start_date = arrow.get(start_date).datetime
+            if (
+                start_date
+                > telescopeAllocationLookup[allocationAndStartDate["allocation_id"]][
+                    "morning"
+                ]
+            ):
+                continue
+            if telescopeAllocationLookup[earliest["allocation_id"]]["morning"] is False:
+                earliest = allocationAndStartDate
+                continue
+            if (
+                telescopeAllocationLookup[allocationAndStartDate["allocation_id"]][
+                    "morning"
+                ]
+                < telescopeAllocationLookup[earliest["allocation_id"]]["morning"]
+            ):
+                earliest = allocation
+                continue
+        allocationByPlanLookup[plan_id] = [earliest]
+
+        # check if that plan is more urgent than the current plan_with_priority
+        if telescopeAllocationLookup[earliest["allocation_id"]]["morning"] is None:
+            continue
+        if plan_with_priority["morning"] is None:
+            plan_with_priority = {
+                "plan_id": plan_id,
+                "morning": telescopeAllocationLookup[earliest["allocation_id"]][
+                    "morning"
+                ],
+                "start_date": earliest["start_date"],
+            }
+            continue
+        if (
+            telescopeAllocationLookup[earliest["allocation_id"]]["morning"]
+            <= plan_with_priority["morning"]
+            and earliest["start_date"] < plan_with_priority["start_date"]
+        ):
+            plan_with_priority = {
+                "plan_id": plan_id,
+                "morning": telescopeAllocationLookup[earliest["allocation_id"]][
+                    "morning"
+                ],
+                "start_date": earliest["start_date"],
+            }
+            continue
+
+    return plan_with_priority["plan_id"]
+
+
 def service(queue):
     while True:
         if len(queue) == 0:
-            time.sleep(1)
+            time.sleep(2)
             continue
 
-        plans, survey_efficiencies, combine_plans, default_plan, user_id = queue[0]
-        plan_ids = []
-
         with DBSession() as session:
+            queue_copy = deepcopy(queue)
+            index = prioritize_queue(queue_copy, session)
```

Comment:
For safety, this could be wrapped in a try except to just return 0 if something went wrong?

Reply:
Yep good idea, didn't add it for testing but needed before merging for sure.